File: apps/api/app/schemas/trades.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field, model_validator

from app.models.enums import AssetClass, ExitReason, StrategyType, TradeEventType, TradeStatus


class TradeCreate(BaseModel):
    watchlist_item_id: int | None = None
    signal_id: int | None = None
    strategy_type: StrategyType | None = None
    entry_price: Decimal = Field(gt=0)
    stop_loss: Decimal = Field(gt=0)
    target_1: Decimal | None = Field(default=None, gt=0)
    target_2: Decimal | None = Field(default=None, gt=0)
    position_size: Decimal = Field(gt=0)
    fees: Decimal | None = Field(default=None, ge=0)
    opened_at: datetime
    notes: str | None = None

    @model_validator(mode="after")
    def validate_long_trade_plan(self) -> "TradeCreate":
        if self.watchlist_item_id is None and self.signal_id is None:
            raise ValueError("watchlist_item_id or signal_id is required.")
        if self.entry_price <= self.stop_loss:
            raise ValueError("entry_price must be greater than stop_loss for long trades.")
        if self.target_1 is not None and self.target_1 <= self.entry_price:
            raise ValueError("target_1 must be greater than entry_price.")
        if self.target_2 is not None and self.target_2 <= self.entry_price:
            raise ValueError("target_2 must be greater than entry_price.")
        return self


class TradeRead(BaseModel):
    id: int
    signal_id: int | None
    watchlist_item_id: int
    status: TradeStatus
    strategy_type: StrategyType
    asset_class: AssetClass
    symbol: str
    entry_price: Decimal
    stop_loss: Decimal
    target_1: Decimal | None
    target_2: Decimal | None
    position_size: Decimal
    fees: Decimal | None
    opened_at: datetime
    closed_at: datetime | None
    exit_price: Decimal | None
    exit_reason: ExitReason | None
    risk_per_unit: Decimal
    initial_risk_amount: Decimal | None
    initial_risk_percent: Decimal | None
    initial_risk_reward: Decimal | None
    target_1_potential_r: Decimal | None
    target_2_potential_r: Decimal | None
    result_amount: Decimal | None
    result_r: Decimal | None
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TradeEventCreate(BaseModel):
    event_type: TradeEventType
    event_time: datetime
    price: Decimal | None = Field(default=None, gt=0)
    reason: str | None = Field(default=None, max_length=255)
    notes: str | None = None

    @model_validator(mode="after")
    def validate_management_event(self) -> "TradeEventCreate":
        if self.event_type not in {
            TradeEventType.NOTE,
            TradeEventType.STOP_UPDATED,
            TradeEventType.TARGET_UPDATED,
        }:
            raise ValueError("Only note, stop update, and target update events are supported.")
        if self.event_type == TradeEventType.NOTE and not self.notes:
            raise ValueError("notes are required for note events.")
        if self.event_type == TradeEventType.STOP_UPDATED and self.price is None:
            raise ValueError("price is required for stop update events.")
        if self.event_type == TradeEventType.TARGET_UPDATED:
            if self.price is None:
                raise ValueError("price is required for target update events.")
            if self.reason not in {"target_1", "target_2"}:
                raise ValueError("reason must be target_1 or target_2 for target updates.")
        return self


class TradeEventRead(BaseModel):
    id: int
    trade_id: int
    event_type: TradeEventType
    event_time: datetime
    price: Decimal | None
    quantity: Decimal | None
    old_value: str | None
    new_value: str | None
    reason: str | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class TradeDetailRead(TradeRead):
    events: list[TradeEventRead]


class TradeClose(BaseModel):
    exit_price: Decimal = Field(gt=0)
    exit_reason: ExitReason
    closed_at: datetime
    notes: str | None = None

File: apps/web/src/types/trades.ts
import type { AssetClass, StrategyType } from "@/types/signals";

export type TradeStatus = "open" | "partial_profit" | "break_even" | "exit_warning" | "exit_signal" | "closed" | "reviewed";

export type ExitReason = "stop_loss" | "target_1" | "target_2" | "manual_exit" | "structure_break" | "time_stop" | "failed_breakout" | "other";

export type TradeEventType = "note" | "stop_updated" | "target_updated" | "closed";

export type TradeEvent = {
  id: number;
  trade_id: number;
  event_type: TradeEventType;
  event_time: string;
  price: string | null;
  quantity: string | null;
  old_value: string | null;
  new_value: string | null;
  reason: string | null;
  notes: string | null;
  created_at: string;
};

export type Trade = {
  id: number;
  signal_id: number | null;
  watchlist_item_id: number;
  status: TradeStatus;
  strategy_type: StrategyType;
  asset_class: AssetClass;
  symbol: string;
  entry_price: string;
  stop_loss: string;
  target_1: string | null;
  target_2: string | null;
  position_size: string;
  fees: string | null;
  opened_at: string;
  closed_at: string | null;
  exit_price: string | null;
  exit_reason: ExitReason | null;
  risk_per_unit: string;
  initial_risk_amount: string | null;
  initial_risk_percent: string | null;
  initial_risk_reward: string | null;
  target_1_potential_r: string | null;
  target_2_potential_r: string | null;
  result_amount: string | null;
  result_r: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type TradeDetail = Trade & {
  events: TradeEvent[];
  journal_entry: JournalEntry | null;
};

export type JournalEntry = {
  id: number;
  trade_id: number;
  user_id: number;
  setup_rule_followed: boolean | null;
  entry_quality_score: number | null;
  stop_quality_score: number | null;
  exit_quality_score: number | null;
  discipline_score: number | null;
  market_context: string | null;
  emotional_notes: string | null;
  what_went_well: string | null;
  what_went_wrong: string | null;
  lesson_learned: string | null;
  reviewed_at: string | null;
  created_at: string;
  updated_at: string;
};

export type TradeCreatePayload = {
  watchlist_item_id?: number;
  signal_id?: number;
  strategy_type?: StrategyType;
  entry_price: string;
  stop_loss: string;
  target_1?: string | null;
  target_2?: string | null;
  position_size: string;
  fees?: string | null;
  opened_at: string;
  notes?: string | null;
};

export type TradeEventCreatePayload = {
  event_type: TradeEventType;
  event_time: string;
  price?: string | null;
  reason?: string | null;
  notes?: string | null;
};

export type TradeClosePayload = {
  exit_price: string;
  exit_reason: ExitReason;
  closed_at: string;
  notes?: string | null;
};

export type JournalEntryCreatePayload = {
  setup_rule_followed?: boolean | null;
  entry_quality_score?: number | null;
  stop_quality_score?: number | null;
  exit_quality_score?: number | null;
  discipline_score?: number | null;
  market_context?: string | null;
  emotional_notes?: string | null;
  what_went_well?: string | null;
  what_went_wrong?: string | null;
  lesson_learned?: string | null;
  reviewed_at: string;
};

export type TradeFilters = {
  opened_from?: string;
  opened_to?: string;
  strategy_type?: StrategyType;
  asset_class?: AssetClass;
  reviewed?: boolean;
  setup_rule_followed?: boolean;
};
